src/storage/aggregator.cc

```diff
@@ -143,48 +156,136 @@ bool Aggregator::Update(const std::string& key, const std::string& row, const ui
+        if (!ok) {
+            PDLOG(ERROR, "GetAggrBufferFromRowView failed");
+            status_.store(AggrStat::kUnInit, std::memory_order_relaxed);
+            return false;
+        }
+        recovery_offset = std::min(recovery_offset, buffer.binlog_offset_);
+        aggr_latest_offset = std::max(aggr_latest_offset, buffer.binlog_offset_);
+        int64_t latest_ts = buffer.ts_end_ + 1;
+        uint64_t latest_binlog = buffer.binlog_offset_ + 1;
+        buffer.clear();
+        buffer.ts_begin_ = latest_ts;
+        buffer.binlog_offset_ = latest_binlog;
+        if (window_type_ == WindowType::kRowsRange) {
+            buffer.ts_end_ = latest_ts + window_size_ - 1;
+        }
+        aggr_buffer_map_.emplace(it->GetPK(), std::move(tmp_buffer));
+        it->NextPK();
+    }
+
+    ::openmldb::log::LogReader log_reader(log_parts, base_replicator_->GetLogPath(), false);
+    log_reader.SetOffset(recovery_offset);
+    ::openmldb::api::LogEntry entry;
+    uint64_t cur_offset = recovery_offset;
+    std::string buffer;
+    int last_log_index = log_reader.GetLogIndex();
+    while (true) {
+        buffer.clear();
+        ::openmldb::base::Slice record;
+        ::openmldb::log::Status status = log_reader.ReadNextRecord(&record, &buffer);
+        if (status.IsWaitRecord()) {
+            int end_log_index = log_reader.GetEndLogIndex();
+            int cur_log_index = log_reader.GetLogIndex();
+            if (end_log_index >= 0 && end_log_index > cur_log_index) {
+                log_reader.RollRLogFile();
+                continue;
+            }
+            break;
+        }
+        if (status.IsEof()) {
+            if (log_reader.GetLogIndex() != last_log_index) {
+                last_log_index = log_reader.GetLogIndex();
+                continue;
+            }
+            break;
+        }
+
+        if (!status.ok()) {
+            PDLOG(WARNING, "read binlog failed: %s", status.ToString().c_str());
+            continue;
+        }
+
+        bool ok = entry.ParseFromString(record.ToString());
+        if (!ok) {
+            PDLOG(WARNING, "parse binlog failed");
+            continue;
+        }
+        if (cur_offset >= entry.log_index()) {
+            continue;
+        }
+
+        if (entry.has_method_type() && entry.method_type() == ::openmldb::api::MethodType::kDelete) {
+            continue;
+        }
+        for (int i = 0; i < entry.dimensions_size(); i++) {
+            const auto& dimension = entry.dimensions(i);
+            if (dimension.idx() == index_pos_) {
+                Update(dimension.key(), entry.value(), entry.log_index(), true);
+                break;
+            }
+        }
+        cur_offset = entry.log_index();
+    }
+    if (cur_offset < aggr_latest_offset) {
+        PDLOG(ERROR, "base table is slower than aggregator");
+        status_.store(AggrStat::kUnInit, std::memory_order_relaxed);
+        return false;
+    }
```

Comment:
need not to recover from snapshot ?

Reply:
> need not to recover from snapshot ?

 binlog is enough for recover?

Reply:
There is a potential issue here. As some old binlog files may be deleted after snapshotting, there may be a scenario where some entries that are required for aggr recover are deleted from binlog (but exists in the snapshot).

Maybe a quickfix is:
- whenever tablet doing `SchedDelBinlog`, we force flushing the aggregators (but still keep the memory states for newer updates). 
- Later flushing triggered by `CheckBufferFilled` will override or update the aggr result in the aggr table.

When we change the aggregator update to be async, we can have a nicer strategy to guarantee this.

What do you think? @dl239 @nautaa